### PR TITLE
Allow Stack to implement Middleware

### PR DESCRIPTION
**Pull request by shadowhand**
Being able to use stacks as middleware allows for complex compositions.

Refs #3

### CODE REVIEW

File: src/StackInterface.php [hunk 1]
```diff
@@ -43,5 +17,5 @@ public function withoutMiddleware(MiddlewareInterface $middleware);
      *
      * @return ResponseInterface
      */
-    public function process(RequestInterface $request);
+    public function dispatch(RequestInterface $request);
```
Comment (harikt) [inline src/StackInterface.php:20]: What is the reason behind the change in method name ?

Comment (mnapoli) [inline src/StackInterface.php:20]: @harikt See the issue.

Comment (harikt) [inline src/StackInterface.php:20]: @mnapoli yes I did go through the groups discussion and issue. Probably I didn't quite understand whether change in method name have any benefit. I will go though it once again to understand the reasoning.

Comment (mnapoli) [inline src/StackInterface.php:20]: Here is for example an implementation of a middleware pipe _as a middleware_: https://github.com/stratifyphp/http/blob/master/src/Middleware/Pipe.php#L36

With the current interface a pipe/stack cannot be a middleware at the same time because the methods have the same name.

I don't know if a lot of projects do that, but this is really great to compose middlewares and create complex architectures, here is a very basic example: https://gist.github.com/mnapoli/4b082c1a9812904cddc1a5719d01e889 (as you can see the router is also a middleware)

Comment (mindplay-dk) [inline src/StackInterface.php:20]: > With the current interface a pipe/stack cannot be a middleware at the same time because the methods have the same name.

That's problematic, I agree.

**Comment by shadowhand** [comment]
@mindplay-dk if you're in favor, can you approve the PR please?

File: src/StackInterface.php (hunk 1 above)
Comment (markstory) [reply src/StackInterface.php:20]: Pipelines as middleware is something we're exploring [for CakePHP](https://github.com/cakephp/cakephp/issues/9089) as well. It has some nice benefits in that pipelines and individual middleware are easily composed together.

**Comment by mindplay-dk** [comment]
@shadowhand with `StackInterface` removed by #12, this PR is no longer relevant, correct?

**Comment by shadowhand** [comment]
@mindplay-dk yes.
